src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts

```diff
@@ -156,6 +184,140 @@ export class CommandDetectionCapability implements ICommandDetectionCapability {
+			}
+			// console.log('output', e.getOutput());
+		});
+
+		this.registerCommandFinishedListener(/.+/, command => {
+			const output = command.getOutput();
+			//TODO: support this for tasks
+			//TODO: narrow this down more
+			// if (command.exitCode === 1) {
+			// TODO: add other possible regexps
+			const match = output?.match(/.*address already in use \d\.\d.\d\.\d:(\d\d\d\d).*/);
+			if (match && match.length === 2 && command.endMarker) {
+				console.log(`address in use for port ${match[1]}`);
+				const actions: IAction[] = [];
+				const labelRun = localize("terminal.freePort", 'Free port');
+				actions.push({
+					class: undefined, tooltip: labelRun, id: 'terminal.freePort', label: labelRun, enabled: true,
+					run: async () => {
+						const procs = await new Promise<string>((resolve, reject) => {
+							cp.exec(`lsof -nP -iTCP -sTCP:LISTEN | grep ${match[1]}`, {}, (err, stdout) => {
+								if (err) {
+									return reject('Problem occurred when listing active processes');
+								}
+								resolve(stdout);
+							});
+						});
+						const procLines = procs.split('\n');
+						if (procLines.length >= 1) {
+							const regex = /\s+(\d+)\s+/;
+							const matches = procLines[0].match(regex);
+							if (matches && matches.length > 1) {
+								console.log('found pid', matches[1]);
+								await new Promise<string>((resolve, reject) => {
+									cp.exec(`kill ${matches[1]}`, {}, (err, stdout) => {
+										if (err) {
+											console.log('error', err);
+											return reject(`Problem occurred when killing the process w id: ${matches[1]}`);
+										}
+										console.log(`killed process ${matches[1]}`);
+										resolve(stdout);
+									});
+								});
+							}
+						}
+					}
+				});
+
+				this._registerContextualDecoration(command.endMarker, actions);
+			}
+			// }
+		});
+	}
+
+	private _registerContextualDecoration(marker: IMarker, actions?: IAction[], x?: number, width?: number, border?: boolean): void {
+		if ('registerDecoration' in this._terminal) {
+			const d = (this._terminal as any).registerDecoration({
+				marker,
+				x,
+				width,
+				layer: 'top',
+				actions
+			});
+			d.onRender((e: HTMLElement) => {
+				if (!this._decorationMarkerIds.has(d.marker.id)) {
+					e.style.border = border ? '1px solid #f00' : '';
+					const inner = document.createElement('div');
+					inner.classList.add('codicon-light-bulb', 'codicon', 'terminal-command-decoration', 'xterm-decoration');
+					inner.style.position = 'absolute';
+					inner.style.bottom = '100%';
+					inner.style.left = '0';
+					inner.style.color = '#ffcc00';
+					e.appendChild(inner);
+					if (actions) {
+						this._decorationMarkerIds.add(d.marker.id);
+						this._onRequestCreateContextMenu.fire({ getAnchor: () => e, getActions: () => actions });
+					}
+				}
+			});
+		}
+	}
+
+	registerCommandFinishedListener(matcher: string | RegExp, callback: (command: ITerminalCommand) => void): IDisposable {
```

Comment:
This sort of thing is fine on CommandDetectionCapability but we should be more thoughtful on an ITerminalInstance API. Something like this:

```ts
interface ITerminalInstance {
  registerContextualAction(options: ITerminalContextualActionOptions): IDisposable;
}

interface ITerminalContextualActionOptions {
  commandLineMatcher: string | RegExp;
  // Allows searching a subset of lines efficiently
  outputMatcher?: ITerminalOutputMatcher;
  // Allow the name to be generated based on capture groups in the output matcher(s)
  actionName: string | DynamicActionName;
  callback: ContextualActionCallback;
}

type DynamicActionName = (commandLineMatch: RegExpMatchArray, outputMatch?: RegExpMatchArray | null) => string;
type ContextualActionCallback = (commandLineMatch: RegExpMatchArray, outputMatch?: RegExpMatchArray | null) => void;

interface ITerminalOutputMatcher {
	lineMatcher: string | RegExp; 
	anchor?: 'top' | 'bottom'
	offset?: number;
	length?: number;
}
```

Reply:
If needed we could also have command and arg matchers so that consumers don't need to each parse the command line if they only apply on a certain argument.